### PR TITLE
Implement Dropbox export (#82)

### PR DESCRIPTION
**Pull request by matheusfaustino**
Hello guys,

First of all, thank you for your work and for this addon. I'm sending my initial implementation of a Dropbox integration, related to the issue #82, to see if you guys are interested in this feature, if it's following the coding standards of the project and if I have to change something in the implementation. 

After this one, I'll work on the Google Drive integration, so I will use it as an experiment to the Google Drive integration.

Thanks for your time.

Best,

EDIT:

Ow, you have to create an app in dropbox and get the app key to test the implementation

### CODE REVIEW

**Comment by tophf** [comment]
Don't use await/async since we support old browsers, use URLSearchParams API, chromeLocal and chromeSync wrappers, find a different way to pass the parameters like messaging because redirecting location.href is weird. Find a way not to use web_accessible_resources. There are some nits more I'll pick on in a few days.

**Comment by matheusfaustino** [comment]
Hi @tophf, thanks for the review. I will fix the things this week and update the PR.
`location.href` is weird, I though in using window.open, but it didn't work, I'll try again. 
Unfortunetely, without `web_accessible_resources` dropbox doesn't redirect back the user, then I can't get the access token from dropbox. 

Which are the browsers the project supports? (I'm going to update the README to help others)

**Comment by prototype99** [comment]
if there are any I've missed feel free to say, however my understanding is it's only chrome, Firefox and opera

**Comment by matheusfaustino** [comment]
Hi @tophf, I dove into `URLSearchParams` API and the way that dropbox uses to pass the parameters isn't supported by the APi, that is why the use a custom function to handle that:

https://github.com/dropbox/dropbox-sdk-js/blob/a88a138c0c3260c3537f30f94b003c1cf64f2fbd/examples/javascript/utils.js 

I tried a lot of way, but without success. Here is a URL to "test":

`moz-extension://c418f0ff-99ea-4a45-9571-3529b786ae18/dropbox-oauth.html#access_token=3Y4Z58CKMaoAAAAAAAA4SalyqKgxb_pg31RmjlDCaTJaKl8MAuUvmS7GZ1vyyE1J&token_type=bearer&uid=40150504&account_id=dbid%3AAADMZqIhjR7zFV8TpyEymvwjdI0jc4c2uRY`

**Comment by tophf** [comment]
It works with anything you feed it, but of course in this case you need to extract the hash part first: `const params = new URLSearchParams(new URL(entireUrl).hash)`

**Comment by tophf** [comment]
As for web_accessible_resource, this is really something I'd like to avoid. Look for an alternative solution e.g. chrome.identity.launchWebAuthFlow etc.

**Comment by tophf** [comment]
No, we don't have a server.

**Comment by matheusfaustino** [comment]
Hello @tophf, thank you very much for the tip about `identity.launchWebAuthFlow` that just worked perfectly. Now, there is not web_accessible_resources. I didn't know about launchWebAuthFlow, I really need to update myself about it. 

If you think there is room for improvement, just tell me, please. 

I think would be good if the user has some feedback when the file is uploading, but I couldn't find a nice way to do that.

**Comment by matheusfaustino** [comment]
I waw thinking in using the current dialog to show that it's uploading the file and then close it automatically when it's done. What do you guys think?

**Comment by tophf** [comment]
Progress reporting is a good thing especially when the style database is 5MB or more. I think we'll need a built-in gzip or zip in a pure js worker.

**Comment by matheusfaustino** [comment]
Yeah, I'm going to work on the progress reporting.

About gzip/zip I think it's very interesting idea. I found some libs that can help: 
- https://stuk.github.io/jszip/
- https://github.com/imaya/zlib.js/blob/develop/README.en.md
- https://gildas-lormeau.github.io/zip.js/core-api.html#zip-writing (claims that already uses workers)

If I'd pick one, I'd choose jszip because the api is cleaner than the others. But we can discuss it better.
If we gonna implement the compression, I think we should do that before this PR, because we won't need to implement backward compatibility with decompression and compression version of it.

**Comment by matheusfaustino** [comment]
I googled about progress bar on dropbox sdk and sadly they don't have anything to track the upload (https://github.com/dropbox/dropbox-sdk-js/issues/157). 
I could try simulate a progress bar by separating the file into severals chunks and upload it individually (filesUploadSession*), but maybe it would increase too much the integration complexity.

**Comment by DanaMW** [comment]
I have been known to be wrong once in awhile but you could do it the old fashioned way and get the file size on start and simply have a small routine do the math yourself. I think dropbox is wrong on one count they have to keep track of how much is uploaded or blocks sent so that might allow you to get a progress indicator. Just thinking out loud.

**Comment by matheusfaustino** [comment]
Hi @DanaMW, thank you for your comment. That was what I said in my last statement about filesUploadSession* functions, maybe it wasn't clear, sorry. 

I don't know if it's worth to create a more complex mechanism to upload a file just for a progress bar. Maybe just a couple of messages + compression would solve the problem better.

**Comment by DanaMW** [comment]
Hi again on my machine using tampermonkey it all happens so fast you might not even get the bar drawn before it was done lol. Nice to meet you and have fun here man you picked a good crew.

**Comment by matheusfaustino** [comment]
Hello guys, finally I had some time to implement the compression and decompression of the styles for the export and also I added some messages in the process. I hope you like.
If you need anything else, just tell me.

**Comment by matheusfaustino** [comment]
Hi @Mottie, no problem, I added it now. Thanks.

**Comment by matheusfaustino** [comment]
To resolve the conflict, do you prefer rebase or merge?

**Comment by tophf** [comment]
You can merge the master, we'll simply squash the PR into one commit later.

**Comment by matheusfaustino** [comment]
Merge done ^^

File: manage/import-export-dropbox.js [hunk 1]
```diff
@@ -0,0 +1,191 @@
...
+
+function requestDropboxAccessToken() {
+  const browserApi = typeof browser === 'undefined' ? chrome : browser;
+  const client = new Dropbox.Dropbox({clientId: DROPBOX_API_KEY});
+  const authUrl = client.getAuthenticationUrl(browserApi.identity.getRedirectURL());
+
+  return browserApi.identity.launchWebAuthFlow({url: authUrl, interactive: true})
+  .then(urlReturned => {
+    const params = new URLSearchParams(new URL(urlReturned).hash.replace('#', '?'));
+
```
Comment (tophf) [inline manage/import-export-dropbox.js]: no need to add blank lines in a single-purpose block like this one and all others below

File: manage/import-export-dropbox.js [hunk 2]
```diff
@@ -0,0 +1,191 @@
...
+$('#sync-dropbox-export').onclick = () => {
+
+  messageProgressBar({title: t('bckpDropboxStyles'), text: t('connectingDropbox')});
+
+  hasDropboxAccessToken().then(token => {
+    if (typeof token === 'undefined') {
+      return requestDropboxAccessToken();
+    }
+
```
Comment (tophf) [inline manage/import-export-dropbox.js]: `return token || requestDropboxAccessToken();`

File: manage/import-export-dropbox.js [hunk 3]
```diff
@@ -0,0 +1,191 @@
...
+  })
+  .then(token => {
+    const client = new Dropbox.Dropbox({
+      clientId: DROPBOX_API_KEY,
+      accessToken: token
+    });
+
+    return client.filesDownload({path: '/' + DROPBOX_FILE})
+     .then(_ => messageBox.confirm(t('overwriteFileExport')))
+    .then(ok => {
```
Comment (tophf) [inline manage/import-export-dropbox.js]: subsequent .then should all have the same 2-space indent, not 1 or 0

File: manage/import-export-dropbox.js [hunk 4]
```diff
@@ -0,0 +1,191 @@
...
+}
+
+function requestDropboxAccessToken() {
+  const browserApi = typeof browser === 'undefined' ? chrome : browser;
+  const client = new Dropbox.Dropbox({clientId: DROPBOX_API_KEY});
+  const authUrl = client.getAuthenticationUrl(browserApi.identity.getRedirectURL());
+
+  return browserApi.identity.launchWebAuthFlow({url: authUrl, interactive: true})
+  .then(urlReturned => {
+    const params = new URLSearchParams(new URL(urlReturned).hash.replace('#', '?'));
```
Comment (tophf) [inline manage/import-export-dropbox.js]: due to a bug in older versions of Chrome we simply strip the character e.g. `'#', ''`

File: manage/import-export-dropbox.js [hunk 5]
```diff
@@ -0,0 +1,191 @@
...
+      /** deletes file if user want to */
+      if (!ok) {
+        return Promise.reject({status: HTTP_STATUS_CANCEL});
+      }
+
+      return client.filesDelete({path: '/' + DROPBOX_FILE});
+    })
+    /** file deleted with success, get styles and create a file */
+    .then(_ => {
+      messageProgressBar({title: t('bckpDropboxStyles'), text: t('gettingStyles') });
```
Comment (tophf) [inline manage/import-export-dropbox.js]: * no need to declare an unused parameter, please use `() => `
* extract t('bckpDropboxStyles') as `title` const

File: manage/import-export-dropbox.js [hunk 6]
```diff
@@ -0,0 +1,191 @@
...
+  .then(token => {
+    const client = new Dropbox.Dropbox({
+      clientId: DROPBOX_API_KEY,
+      accessToken: token
+    });
+
+    return client.filesDownload({path: '/' + DROPBOX_FILE})
+     .then(_ => messageBox.confirm(t('overwriteFileExport')))
+    .then(ok => {
+      /** deletes file if user want to */
```
Comment (tophf) [inline manage/import-export-dropbox.js]: `/**` is a jsdoc comment style for describing types of variables/parameters/methods;
please use `//` for single line comments

File: _locales/en/messages.json [hunk 7]
```diff
@@ -1182,10 +1182,43 @@
...
   },
   "bckpInstStyles": {
-      "message": "Export styles"
+    "message": "Export styles"
   },
   "retrieveBckp": {
-      "message": "Import styles"
+    "message": "Import styles"
+  },
+  "bckpDropboxStyles": {
```
Comment (tophf) [inline _locales/en/messages.json]: whoever added `bckp` in our messages was wrong so let's not repeat the mistake: please use `backup` in the new messages. I also think it might be a good idea to prefix all the new messages with something like uhm... `sync` maybe?

File: _locales/en/messages.json [hunk 8]
```diff
@@ -1182,10 +1182,43 @@
...
+  "connectingDropbox": {
+      "message": "Connecting Dropbox."
+  },
+  "gettingStyles": {
+      "message": "Getting all styles."
+  },
+  "compactStyles": {
+      "message": "Compacting styles."
+  },
+  "descompactStyles": {
```
Comment (tophf) [inline _locales/en/messages.json]: no such word in English, also maybe let's use "packing" and "unpacking"? or "zipping" and "unzipping"?

File: _locales/en/messages.json [hunk 9]
```diff
@@ -1182,10 +1182,43 @@
...
+      "message": "Descompating styles."
+  },
+  "readingStyles": {
+      "message": "Reading styles."
+  },
+  "uploadingFile": {
+      "message": "Uploading File."
```
Comment (tophf) [inline _locales/en/messages.json]: no need for the Title Case

File: _locales/en/messages.json [hunk 10]
```diff
@@ -1182,10 +1182,43 @@
...
+    "message": "Do you want to overwrite an existing file?"
+  },
+  "exportSavedSuccess": {
+    "message": "File saved with success"
+  },
+  "noFileToImport": {
+    "message": "You don't have a file to import."
+  },
+  "connectingDropbox": {
+      "message": "Connecting Dropbox."
```
Comment (tophf) [inline _locales/en/messages.json]: all ongoing status messages should probably end with `...`

File: manage/import-export-dropbox.js [hunk 11]
```diff
@@ -0,0 +1,191 @@
...
+    return token;
+  })
+  .then(token => {
+    const client = new Dropbox.Dropbox({
+      clientId: DROPBOX_API_KEY,
+      accessToken: token
+    });
+
+    return client.filesDownload({path: '/' + DROPBOX_FILE})
+     .then(_ => messageBox.confirm(t('overwriteFileExport')))
```
Comment (tophf) [inline manage/import-export-dropbox.js]: no need to declare an unused parameter, please use `() => `

File: manage/compress-text.js [hunk 12]
```diff
@@ -0,0 +1,35 @@
+/* global messageBox */
+'use strict';
+
+onDOMready().then(_ => {
+  zip.workerScriptsPath = "/vendor/zipjs/";
+});
+
+function createZipFileFromText(filename, text) {
+  return new Promise((resolve, reject) => {
+    // use a BlobWriter to store the zip into a Blob object
```
Comment (tophf) [inline manage/compress-text.js]: no need for the comments in this function

File: manage/compress-text.js [hunk 13]
```diff
@@ -0,0 +1,35 @@
...
+    zip.createWriter(new zip.BlobWriter('application/zip'), writer => {
+      // use a TextReader to read the String to add
+      writer.add(filename, new zip.TextReader(text), function () {
+        // close the zip writer
+        writer.close(blob => {
+          // blob contains the zip file as a Blob object
+          resolve(blob);
+        });
+      });
+    }, error => reject(error));
```
Comment (tophf) [inline manage/compress-text.js]: `}, reject);`

File: manage/compress-text.js [hunk 14]
```diff
@@ -0,0 +1,35 @@
...
+function readZipFileFromBlob(blob) {
+  return new Promise((resolve, reject) => {
+    zip.createReader(new zip.BlobReader(blob), zipReader => {
+      zipReader.getEntries(entries => {
+        entries[0].getData(new zip.BlobWriter('text/plain'), data => {
+          zipReader.close();
+          resolve(data);
+        });
+      });
+    }, error => reject(error))
```
Comment (tophf) [inline manage/compress-text.js]: `}, reject);`

**Comment by matheusfaustino** [comment]
I'm pretty sure I fixed the eslint error... that's weird. Thanks for the feedback.

File: manage.html [hunk 15]
```diff
@@ -161,22 +161,28 @@ <h2 class="style-name">
...
   <script src="manage/import-export.js" async></script>
+
   <script src="manage/incremental-search.js" async></script>
   <script src="msgbox/msgbox.js" async></script>
   <script src="js/sections-equal.js" async></script>
   <script src="js/storage-util.js" async></script>
+
+  <script src="vendor/dropbox/dropbox-sdk.js" async></script>
+  <script src="vendor/zipjs/zip.js" defer></script>
+  <script src="manage/compress-text.js" defer></script>
```
[the remark this answers is not in the document]
Comment (matheusfaustino) [reply manage.html]: root or inside another one?

[the remark this answers is not in the document]
Comment (tophf) [reply manage.html]: I think it should be in the root folder because we might use the sync stuff from elsewhere.

File: _locales/en/messages.json (hunk 9 above)
Comment (matheusfaustino) [reply _locales/en/messages.json]: Hi, I didn't get this one. Should I remove the Title Case from the "uploadingFile" and "readingStyles" key and keep all of them lower case? Thanks for your time and patience.

**Comment by matheusfaustino** [comment]
Sorry for the long time, I'm going to be more active to deliver this feature.

File: _locales/en/messages.json (hunk 9 above)
Comment (tophf) [reply _locales/en/messages.json]: "Uploading file..."

Comment (matheusfaustino) [reply _locales/en/messages.json]: Ok, I fixed that with the last commit ^^

**Comment by matheusfaustino** [comment]
I fixed the last eslint error from the branch ^^

**Comment by matheusfaustino** [comment]
Solved conflict with the `_locales/en/messages.json`

**Comment by tophf** [comment]
I thought someone would test this feature but no one did. I've just tried clicking the "Dropbox export" button in Chrome and it failed with an error:

```
dropbox-sdk.js:4148 Uncaught (in promise) Error: A client id is required. You can set the client id using .setClientId().
    at Dropbox.getAuthenticationUrl (dropbox-sdk.js:4148)
    at requestDropboxAccessToken (import-export-dropbox.js:33)
    at hasDropboxAccessToken.then.token (import-export-dropbox.js:52)
```

I also have doubts about the export/import buttons - we have too many of them...

**Comment by narcolepticinsomniac** [comment]
I'm not the target demo here, as I'm unlikely to use sync, but it's always a shame when a feature that so many users say they want, gets so far along and goes nowhere, so I figured I'd test it.

I don't get it tbh, since it doesn't seem to actually do anything. There only appears to be the two buttons, with no other fields to fill in, or any kind of explanation, so I assume it's supposed to be a pretty automated process, but clicking the buttons doesn't accomplish much. I get a "connecting to Dropbox" message and the same error tophf mentioned.

I also agree that the buttons aren't ideal. At the very least, their message should avoid "import/export" in favor of "sync" messaging to avoid confusion. Maybe even use icons instead to really differentiate, if that can be pulled off decently.

**Comment by DanaMW** [comment]
it appears he is attempting to us the oauth from a previous web connection

**Comment by narcolepticinsomniac** [comment]
Something similar to these might work:

Header/title:

Dropbox icon [here](https://simpleicons.org/).

Import:

https://octicons.github.com/icon/cloud-download/

Export:

https://octicons.github.com/icon/cloud-upload/

**Comment by DanaMW** [comment]
couldn't we develop a web login that returns the key similar to how say normal Linus does it I figure there is 200 thousand people out there that would give up on it in this form. Like a web log in? never looked at it but I'll get to paying more attention here now.

**Comment by matheusfaustino** [comment]
Hi @tophf , about the buttons I agree with you. I think it's too many, imagine when we added the Google Drive integration.

**Comment by DanaMW** [comment]
yakyak here on GitHub has a an integration that you might look over...

**Comment by narcolepticinsomniac** [comment]
>Hi, I'm sorry, maybe I wasn't clear about how test it, in my first message.

You probably were, and I just didn't read it. Not just because I'm a lazy schmuck either, but to test how I would evaluate the feature if I encountered it as it is. 

The steps you're referring to now aren't realistic IMO. It needs to be simpler to figure out and use.

**Comment by matheusfaustino** [comment]
Hi @DanaMW , what I found about others extensions doing the dropbox integration is they generally have its own dropbox app they use that key. So, it'll be a unique key for all users.

**Comment by DanaMW** [comment]
When you  log in with web it sets your key locally I think  extensions  can read cookies and that's what would have to happen

**Comment by k1nghat** [comment]
this is how syncmarx does dropbox auth but the code isnt public yet:
image

**Comment by matheusfaustino** [comment]
I don't think we can manipulate the dropbox files from the user without an app. 
@DanaMW are you referring to the dropbox key? Sorry, I didn't get :p 

look how the file is kept using dropbox: `Dropbox > Applications > Stylus`

look how the file is kept using dropbox

**Comment by matheusfaustino** [comment]
k1nghat I didn't know this app. I'll look at it, thanks. However, I think if we have a dropbox app, it will be pretty straightforward to the user, it's basically a confirmation popup (if it's logged in dropbox). 

The redirect uri doesn't change in a live extension: https://developer.chrome.com/apps/identity#method-getRedirectURL

**Comment by narcolepticinsomniac** [comment]
Check out how xStyle does it, as it behaves more like I'd expect. When you click the button, it opens a dropbox API permission dialogue:

xstyle

**Comment by matheusfaustino** [comment]
Well, that is what it happens when you use a dropbox. We just need an official dropbox app for Stylus and it's all done and it goes for syncmarx too.

**Comment by matheusfaustino** [comment]
Do you know if Stylus has an official account in dropbox? We can create the app and test with the "real" app.

**Comment by DanaMW** [comment]
@narcolepticinsomniac  that's how most do it that is what I am talking about

**Comment by DanaMW** [comment]
YakYak uses an "iPhone" type  that's why i mentioned it it is to Google for hangouts but same principle applies

**Comment by DanaMW** [comment]
don't actually have to have an app license. Is there a cost to get these guys one because i would put up for one or two ...

**Comment by narcolepticinsomniac** [comment]
>We just need an official dropbox app for Stylus and it's all done and it goes for syncmarx too.

Forgive the ignorance, but if users are syncing through personal accounts, why would Stylus need an "official" account?

If it's just a formality, and all that is required is setting up a free account, then make one. If we have to pay for the privilege, that's something we'd need to discuss, even if big Daddy Warbucks @DanaMW is volunteering.

**Comment by DanaMW** [comment]
Not paying 20 per month for each 4 users lol can't see anything reasonable sooo

**Comment by DanaMW** [comment]
right JUST need to access cookies and each person would auth themselves He has a set way and seems locked in on it

**Comment by DanaMW** [comment]
so ill finish what I'm doing and ill see if i can learn anything helpful this afternoon.

**Comment by DanaMW** [comment]
```
2. Regardless of the type of app, the redirect URI would be the URI where you want to send the user back to after they approve your app to access their account. I can't offer guidance on how to build a Chrome extension in particular, but I know Chrome extensions sometimes use "chrome-extension://" URLs for their option pages, so you may want to look into that. I recommend reading the OAuth guide for context on how the flow works.

 

(Also, if you use the "code" flow, the redirect URI isn't strictly required. If you omit it, the user just needs to copy/paste the authorization code into your app.)

 

4. No, the user does not need to register their own API app. Your app uses its own app key to send them though the OAuth app authorization flow. I recommend trying the example to see how it works. 

 

Production status just removes the limit on the number of users that can connect to your app.

 

In your code, you're attempting to make a GET request to the /oauth2/authorize page. You instead need to send the user to that page as a normal navigation operation. E.g., have them click on the link as shown in the example, or redirect them via JavaScript, such as using window.location.replace(url) or window.location.href.
```

**Comment by DanaMW** [comment]
https://www.dropbox.com/developers/reference/developer-guide#production-approval

**Comment by matheusfaustino** [comment]
@narcolepticinsomniac  because with an official app you can set icon, have a specific folder inside user account and just require the user click "allow", basically. And the others extensions do the same, all of them have a dropbox app to integrate with dropbox. 

The tough part is testing it, because the redirect uri changes every time. 

I changed my app for production and they shown me basically this https://www.dropbox.com/developers/reference/developer-guide#production-approval .

We don't have to pay for an account.

**Comment by narcolepticinsomniac** [comment]
>We don't have to pay for an account.

SGTM. Are you asking one of us to do so?

>The tough part is testing it, because the redirect uri changes every time.

IDK what that means really, so would making an account resolve this issue?

**Comment by narcolepticinsomniac** [comment]
If it's as simple as needing to set up a free account, I guess I can do that. If there's any extra configuration needed, you can either walk me through it, or we can give you access to it while you work on it.

**Comment by matheusfaustino** [comment]
@narcolepticinsomniac it will probably solve the problem.
Could one of you create an account, please?

**Comment by matheusfaustino** [comment]
@narcolepticinsomniac of course I can help you ^^

**Comment by DanaMW** [comment]
That's @N that would be best . It's a "Development" account 50 testers max till you test

**Comment by DanaMW** [comment]
@tophf  would prefer we use the Chat server and not fill the channels so if woulcould work it out there we would  take stress off  an hard working guy BTW, you do as you like  i am just relaying what he has asked the rest to do... need help? I'm down and i am onboard for testing my dropbox i will test with is empty so we can't hurt it

**Comment by DanaMW** [comment]
well not empty , it has a "fuckdropbox" file in there lol

**Comment by narcolepticinsomniac** [comment]
Made a free account. @matheusfaustino Do you need access to it while you work on it? If you're on Discord, I can DM you the account info.

**Comment by matheusfaustino** [comment]
@narcolepticinsomniac nice, I can setup with this account to be easier to test. My discord account is matheuspfaustino#9331 . Thank you very much

**Comment by narcolepticinsomniac** [comment]
We don't share a server, so I guess you need to accept my friend request to receive my messages. You should probably join our server anyway.

**Comment by narcolepticinsomniac** [comment]
https://discord.gg/tzXVNJz

**Comment by matheusfaustino** [comment]
Done ^^ i'm in the server and accepted the request.

**Comment by matheusfaustino** [comment]
Hi guys, I setup the dropbox app for Stylus to make it easy to test. Now, I'm using the official account and app key.

Sadly, to test it, you should you Firefox, because with chrome I couldn't fixed the addon's id. The problem of not fixing the id is that every different installation, of the addon, on chrome will result in a different ID, so a different redirect url to add on dropbox for testing. With Firefox, it's possible to fixed the id as well as the redirect url, so it will be same for everyone.

TL;DR : For now, use firefox for testing ([web-ext](https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/Getting_started_with_web-ext)), please. Meanwhile, I'll try to workaround the chrome situation.

Another topic, @tophf raised the question about too many buttons. What do you think of making the export/import buttons a dropdown ([example](https://getbootstrap.com/docs/4.0/components/dropdowns/))? Using a dropdown, we can add integrations without polluting the UI too much.

File: manifest.json [hunk 16]
```diff
@@ -5,6 +5,12 @@
   "description": "__MSG_description__",
   "homepage_url": "https://add0n.com/stylus.html",
   "manifest_version": 2,
+  "applications": {
+    "gecko": {
+      "id": "{a9449967-86c7-4d0e-8285-322cef7c40f2}",
+      "strict_min_version": "42.0"
+    }
+  },
   "icons": {
```
Comment (Mottie) [inline manifest.json]: Please remove the "applications" section from the manifest... I can't load the extension in chrome with it there.

**Comment by DanaMW** [comment]
Maybe  dropdown of buttons keeping in mind you will  be adding google so maybe consider that when designing it

File: manifest.json (hunk 16 above)
Comment (matheusfaustino) [reply manifest.json]: It's there for testing the integration, I'm going to remove it later :p  

That is weird... I can load the extension on chrome even with this section, it just shows me a warning, but it works. 

Chrome version: Version 68.0.3440.106 (Official Build) (64-bit)

**Comment by narcolepticinsomniac** [comment]
>I can't load the extension in chrome with it there.

Didn't he just tell us it doesn't work in Chromium? I appreciate the effort, but to me, if it doesn't work in Chromium, it doesn't work.

**Comment by Mottie** [comment]
> Didn't he just tell us it doesn't work in Chromium

I posted that right after the update and before I refreshed the page in GitHub... so I missed the comment.

**Comment by narcolepticinsomniac** [comment]
Well, I doubt I have much use for the feature anyway, but I practically never use anything but side-loaded versions, so it'd be useless anyway.

IDK if that'll be a dealbreaker for tophf. Probably, if it's a feature he ever intends to use.

**Comment by Mottie** [comment]
Ok, I see... it is providing the same JSON file, except that it's named `stylus.json` and compressed into a zip file.

**Comment by matheusfaustino** [comment]
@Mottie yeap, @tophf told me to compress it in order to be light to download and upload.

**Comment by Mottie** [comment]
Well, it's working great for me. I can understand your frustration with how Dropbox is so restrictive.

I think it might be nice to be able to choose different backup files from Dropbox. Say, I'm switching between sets of styles for testing, or something.

**Comment by matheusfaustino** [comment]
Thanks ^^, It's hard to explain why it doesn't work in some cases. 

That is a interesting idea, I'll think about it. In order to work this way, we should ask for a filename (upload) and list the files to choose (download).

**Comment by Mottie** [comment]
What is we have one set of import/export buttons and a select next to it?

2018-09-03 20_57_22-stylus

File: manifest.json (hunk 16 above)
Comment (Mottie) [reply manifest.json]: Ahh, you're right. It still works, but there is an error message

2018-09-03 21_05_46-extensions

**Comment by matheusfaustino** [comment]
I was thinking something more like this: https://jsfiddle.net/v0x891do/ 
but the select it's a good option.

**Comment by Mottie** [comment]
So what's the consensus on the import/export buttons?

**Comment by Mottie** [comment]
Errr wait LOL... I exported an empty file 😛

**Comment by matheusfaustino** [comment]
:p whew... that scared me hahaha

**Comment by Mottie** [comment]
Ok, everything works!

So, about the dropdowns... I like the changes, but the buttons don't match the other dropdowns we have on the page. Instead of adding `<i class="arrow down"></i>`, add:

```html
<svg class="svg-icon select-arrow"><use xlink:href="#svg-icon-select-arrow"/></svg>
```

**Comment by matheusfaustino** [comment]
Done ^^ thanks @Mottie 

One thing that we have to prepare is the url on dropbox. 
Is your addon ID on chrome "clngdbkpkpeebahjckkjfobafhncgmne"? 
- it will be: https://clngdbkpkpeebahjckkjfobafhncgmne.chromiumapp.org/ ([see](https://developer.chrome.com/apps/identity#method-getRedirectURL))

And the addon ID on Firefox "{7a7a4a92-a2a0-41d1-9fd7-1e92480d612d}"?
- on MDN we don't have a specification for the url, but after the update we can just get the url generated and save it on dropbox ([see](https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/API/identity#Getting_the_redirect_URL))

**Comment by Mottie** [comment]
Yes

* "clngdbkpkpeebahjckkjfobafhncgmne" is the ID of the extension when installed from the CWS and also the Opera addons page. @tophf I'm not sure who is updating the Opera addon site, but it's currently on v1.4.12.
* "19f1eea9-d818-427e-bdfc-e48d4e08c83c" is the ID of the Firefox extension.

Before we can merge, you'll need to remove the "applications" entry in the manifest.json and rebase the branch.

**Comment by matheusfaustino** [comment]
hi @tophf , actually there is a way to test on chrome with a random id. We just have to discovery the redirect url from chrome to the temp extension and add it on dropbox. 

If the url isn't on dropbox, using Firefox, it'll open a window showing the url and the error. On chrome, that window won't appear, so just put an alert/log [here](https://github.com/openstyles/stylus/pull/393/files#diff-1ae9f80df58b5de48eec7f38b051b317R9) and send it to me, then I can add your url on dropbox and that is it (for locally loaded extension). I'll be online for a couple of hours, if you wanna try it.

File: sync/import-export-dropbox.js [hunk 17]
```diff
@@ -0,0 +1,179 @@
...
+  messageProgressBar({title: title, text: t('connectingDropbox')});
+
+  hasDropboxAccessToken()
+  .then(token => token || requestDropboxAccessToken())
+  .then(token => {
+    const client = new Dropbox.Dropbox({
+      clientId: DROPBOX_API_KEY,
+      accessToken: token
+    });
+
```
Comment (tophf) [inline sync/import-export-dropbox.js]: These empty lines before `return` visually regroup the code in the wrong way - separating each .then block in the middle. Please remove them everywhere.

Comment (matheusfaustino) [reply sync/import-export-dropbox.js]: ok ^^

Comment (matheusfaustino) [reply sync/import-export-dropbox.js]: done :)

**Comment by Mottie** [comment]
Please rebase with the master... there are conflicts, so we can't just push the button 😉

**Comment by narcolepticinsomniac** [comment]
I see no conflicts anymore, but yeah, better to wait for tophf to give it the green light.

**Comment by Kenya-West** [comment]
@matheusfaustino please, make the same for OneDrive!

**Comment by matheusfaustino** [comment]
Hi @Kenya-West , I'll look into the documentation to see how hard is to implement OneDrive, I personally don't use it, but if it's not that hard, why not?

**Comment by Kenya-West** [comment]
> Hi @Kenya-West , I'll look into the documentation to see how hard is to implement OneDrive, I personally don't use it, but if it's not that hard, why not?

I think, OneDrive API is included in Microsoft Graph API, see [graph.microsoft.com](https://graph.microsoft.com) or [dev.onedrive.com](https://dev.onedrive.com). I personally didn't work with Microsoft Graph, but I know, that Microsoft Graph is supporting both **v1** and **v2** versions. **v1** has more APIs specific for Personal accounts and Work accounts, while **v2** unifies Personal and Work accounts and offers single API queries, no matter business, school or personal subscription user is on. Despite **v2** has a bit less functionality, it is very common action to upload and download files from/to specific folder in OneDrive, so this API method should exist in **v2**.

I suggest you to look at Tampermonkey's source code if possible. and probably [Office Online](https://chrome.google.com/webstore/detail/office-online/ndjpnladcallmjemlbaebfadecfhkepb) add-on.

I will also investigate deep into this, but with no guarantees - I don't have experience in developing browser extensions (but I developed some Angular components just for run).

**Comment by matheusfaustino** [comment]
Hi @tophf , how are you? I hope you're doing well.

Do you need something else from me? bye glhf

**Comment by mikhoul** [comment]
[**@matheusfaustino**](https://github.com/matheusfaustino) commented on [1 oct. 2018 à 08:34 UTC−4](https://github.com/openstyles/stylus/pull/393#issuecomment-425890902 "2018-10-01T12:34:11Z - Replied by Github Reply Comments"):
> Hi [@tophf](https://github.com/tophf) , how are you? I hope you're doing well.
> 
> Do you need something else from me? bye glhf

I really hope it will be  merged soon :smile:

**Comment by tophf** [comment]
I got tired so in the foreseeable time span I'm not working on Stylus except maybe on some trivial occasional stuff like CSSLint/parserlib.

**Comment by Mottie** [comment]
Yeah, I'll take care of it soon... I've been busy with work.

**Comment by matheusfaustino** [comment]
Hello, don't worry. I don't want to give trouble, it's just to know which step we are ^^ if you need something, please ping me :)

**Comment by narcolepticinsomniac** [comment]
@matheusfaustino Are you working on this again right now?

**Comment by matheusfaustino** [comment]
@narcolepticinsomniac I'm about to work on it. Actually, I'm testing the integration with OneDrive. 
Yesterday, I fixed the url of Firefox's addon in dropbox settings

**Comment by narcolepticinsomniac** [comment]
In the future, just give me a heads up on Discord or here if you're gonna log in to the account. I got a security alert, and you didn't respond to the ping, so I changed the password. I can DM you the new one if you need it.

**Comment by matheusfaustino** [comment]
I'm sorry, I took the liberty of doing that, my bad. Ok, if I need it, I ping you. thanks ^^

**Comment by narcolepticinsomniac** [comment]
It's all good. Mottie's got it too if I'm not around.

**Comment by narcolepticinsomniac** [comment]
@matheusfaustino Check Discord.

**Comment by Mottie** [comment]
I'm working on the issue now... essentially, Dropbox has locked our app because we never applied to put our Dropbox app into "production" after we reached 50 users. I'm filling out the form now.

**Comment by matheusfaustino** [comment]
Hi @narcolepticinsomniac sorry I had some problems this weekend. I going to download the discord app to be in touch ^^
Thanks @Mottie , I'm working on that tutorial to test the dropbox implementation.
